File: python/src/scon/resolver.py
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

from .error import SconError, Span
from .format import format_value
from .parser import (
    AstArray,
    AstArraySpread,
    AstArrayValue,
    AstBool,
    AstField,
    AstInclude,
    AstNull,
    AstNumber,
    AstObject,
    AstObjectSpread,
    AstObjectValue,
    AstString,
    AstSubstitution,
    StringInterpolationPart,
    StringLiteralPart,
    parse_document,
)
from .value import SconArray, SconNumber, SconObject, SconValue, clone_value


@dataclass
class LoadOptions:
    include_root: Path | None = None
    max_file_size: int = 16 * 1024 * 1024
    max_include_depth: int = 64
    max_include_files: int = 1024
    max_array_length: int = 1_000_000
    max_object_depth: int = 512


@dataclass
class EvalEntry:
    value: SconValue | EvalObject
    layer: str
    kind: str


class EvalObject:
    def __init__(self) -> None:
        self.entries: dict[str, EvalEntry] = {}


def parse_string(source: str) -> SconValue:
    return Resolver(LoadOptions()).eval(parse_document(source))


def parse_file(path: str | Path, *, include_root: str | Path | None = None) -> SconValue:
    file = Path(path).resolve()
    root = Path(include_root).resolve() if include_root is not None else file.parent
    source = _read_text_raw(file)
    options = LoadOptions(include_root=root)
    if len(source.encode()) > options.max_file_size:
        raise SconError("ResourceLimitExceeded", "maximum file size exceeded")
    resolver = Resolver(options)
    resolver.stack.append(file)
    resolver.seen.add(file)
    return resolver.eval(parse_document(source, str(file)))


def get_path(value: SconValue, path: str) -> SconValue:
    current = value
    for segment in path.split("."):
        if not isinstance(current, SconObject):
            raise SconError("TypeMismatch", "path segment requires object")
        if segment not in current:
            raise SconError("MissingReference", "path is not defined")
        current = current[segment]
    return current


class Resolver:
    def __init__(self, options: LoadOptions) -> None:
        self.options = options
        self.stack: list[Path] = []
        self.seen: set[Path] = set()
        self.cache = {}
        self.root = EvalObject()
        self.in_progress: list[list[str]] = [[]]

    def eval(self, document) -> SconValue:
        self._eval_object(document.root, [], document.file)
        return _public_object(self.root)

    def _eval_object(self, obj: AstObject, path: list[str], file: str | None) -> None:
        if len(path) > self.options.max_object_depth:
            raise SconError("ResourceLimitExceeded", "maximum object depth exceeded", obj.span)
        local_seen = False
        for member in obj.members:
            if isinstance(member, AstObjectSpread):
                if local_seen:
                    raise SconError("InvalidSpread", "object spread must appear before local members", member.span)
                target = self._lookup(member.sub.path, member.span).value
                if not isinstance(target, EvalObject):
                    raise SconError("TypeMismatch", "object spread target is not an object", member.span)
                _overlay_base(self._object_at(path, member.span), target)
            elif isinstance(member, AstInclude):
                included = self._load_include(file, member)
                self._eval_object(included.root, path, included.file)
            else:
                local_seen = True
                self._eval_field(member, path, file)

    def _eval_field(self, field: AstField, current: list[str], file: str | None) -> None:
        target = [*current, *[segment.value for segment in field.path.segments]]
        if isinstance(field.value, AstObjectValue):
            self._ensure_object(target, field.span)
            self.in_progress.append(target)
            try:
                self._eval_object(field.value.object, target, file)
            finally:
                self.in_progress.pop()
            return
        value = self._eval_value(field.value, file)
        self._insert(target, value, "ordinary", field.span)

    def _eval_value(self, value, file: str | None) -> SconValue | EvalObject:
        if isinstance(value, AstNull):
            return None
        if isinstance(value, AstBool):
            return value.value
        if isinstance(value, AstNumber):
            try:
                return SconNumber.parse(value.raw)
            except SconError as exc:
                raise SconError(exc.code, exc.message, value.span) from exc
        if isinstance(value, AstString):
            return self._eval_string(value)
        if isinstance(value, AstSubstitution):
            return _clone_any(self._lookup(value.path, value.span).value)
        if isinstance(value, AstArray):
            out = SconArray()
            self.in_progress.append([])  # prevents direct self-completed array lookup via lookup path check below.
            self.in_progress.pop()
            for item in value.items:
                if len(out) >= self.options.max_array_length:
                    raise SconError("ResourceLimitExceeded", "maximum array length exceeded", item.span)
                if isinstance(item, AstArrayValue):
                    out.append(_public_maybe(self._eval_value(item.value, file)))
                elif isinstance(item, AstArraySpread):
                    target = self._lookup(item.sub.path, item.span).value
                    if not isinstance(target, SconArray):
                        raise SconError("TypeMismatch", "array spread target is not an array", item.span)
                    out.extend(clone_value(v) for v in target)
            return out
        if isinstance(value, AstObjectValue):
            nested = Resolver(self.options)
            nested.stack = self.stack
            nested.seen = self.seen
            nested.cache = self.cache
            nested._eval_object(value.object, [], file)
            return nested.root
        raise SconError("UnexpectedToken", "unknown value")

    def _eval_string(self, value: AstString) -> str:
        if len(value.parts) == 1 and isinstance(value.parts[0], StringLiteralPart):
            return value.parts[0].value
        out = ""
        for part in value.parts:
            if isinstance(part, StringLiteralPart):
                out += part.value
            elif isinstance(part, StringInterpolationPart):
                replacement = self._lookup(part.path, part.span).value
                if isinstance(replacement, str):
                    out += replacement
                elif isinstance(replacement, bool):
                    out += "true" if replacement else "false"
                elif isinstance(replacement, SconNumber):
                    out += replacement.to_text()
                else:
                    raise SconError("TypeMismatch", "interpolation requires string, number, or boolean", part.span)
        return out

    def _lookup(self, path, span: Span) -> EvalEntry:
        names = [segment.value for segment in path.segments]
        if any(active == names for active in self.in_progress):
            raise SconError("MissingReference", "reference is not completed yet", span)
        obj = self.root
        entry = None
        for index, name in enumerate(names):
            entry = obj.entries.get(name)
            if entry is None:
                raise SconError("MissingReference", f"missing reference '{name}'", span)
            if index < len(names) - 1:
                if not isinstance(entry.value, EvalObject):
                    raise SconError("TypeMismatch", "reference path crosses non-object value", span)
                obj = entry.value
        assert entry is not None
        return entry

    def _ensure_object(self, path: list[str], span: Span) -> None:
        obj = self.root
        for index, name in enumerate(path):
            entry = obj.entries.get(name)
            if entry is None:
                child = EvalObject()
                obj.entries[name] = EvalEntry(child, "local", "structural")
                obj = child
                continue
            if not isinstance(entry.value, EvalObject):
                raise SconError("PathConflict", "path conflicts with scalar value", span)
            if index == len(path) - 1 and entry.layer == "local" and entry.kind != "structural":
                raise SconError("PathConflict", "object field conflicts with ordinary value", span)
            entry.layer = "local"
            entry.kind = "structural"
            obj = entry.value

    def _insert(self, path: list[str], value: SconValue | EvalObject, kind: str, span: Span) -> None:
        obj = self.root
        for name in path[:-1]:
            entry = obj.entries.get(name)
            if entry is None:
                child = EvalObject()
                obj.entries[name] = EvalEntry(child, "local", "structural")
                obj = child
            else:
                if not isinstance(entry.value, EvalObject):
                    raise SconError("PathConflict", "path conflicts with scalar value", span)
                obj = entry.value
        leaf = path[-1]
        existing = obj.entries.get(leaf)
        if existing is None:
            obj.entries[leaf] = EvalEntry(value, "local", kind)
            return
        if existing.layer == "base":
            _overlay_local(existing, value, kind)
            return
        raise SconError("DuplicateKey", f"duplicate key '{leaf}'", span)

    def _object_at(self, path: list[str], span: Span) -> EvalObject:
        obj = self.root
        for name in path:
            entry = obj.entries.get(name)
            if entry is None:
                raise SconError("PathConflict", "target object does not exist", span)
            if not isinstance(entry.value, EvalObject):
                raise SconError("PathConflict", "target path is not an object", span)
            obj = entry.value
        return obj

    def _load_include(self, file: str | None, include: AstInclude):
        path = include.path.value
        if _invalid_include_path(path):
            raise SconError("InvalidIncludePath", "invalid include path", include.span)
        root = (self.options.include_root or (Path(file).parent if file else Path("."))).resolve()
        base = Path(file).parent if file else root
        candidate = (base / path).resolve()
        if not _within_root(candidate, root):
            raise SconError("IncludePathDenied", "include path escapes include root", include.span)
        if candidate in self.stack:
            raise SconError("IncludeCycle", f"include cycle: {candidate}", include.span)
        if len(self.stack) >= self.options.max_include_depth:
            raise SconError("ResourceLimitExceeded", "maximum include depth exceeded", include.span)
        self.seen.add(candidate)
        if len(self.seen) > self.options.max_include_files:
            raise SconError("ResourceLimitExceeded", "maximum include file count exceeded", include.span)
        if candidate in self.cache:
            return self.cache[candidate]
        if not candidate.exists():
            raise SconError("IncludeNotFound", f"include file not found: {candidate}", include.span)
        if not candidate.is_file():
            raise SconError("IncludeNotFile", "include path is not a file", include.span)
        if candidate.stat().st_size > self.options.max_file_size:
            raise SconError("ResourceLimitExceeded", "maximum file size exceeded", include.span)
        self.stack.append(candidate)
        try:
            doc = parse_document(_read_text_raw(candidate), str(candidate))
            self.cache[candidate] = doc
            return doc
        except SconError as exc:
            code = "IncludeRootTypeError" if exc.code == "InvalidRootType" else "IncludeParseError"
            raise SconError(code, exc.message, exc.span) from exc
        finally:
            self.stack.pop()


def _public_object(obj: EvalObject) -> SconObject:
    out = SconObject()
    for key, entry in obj.entries.items():
        out[key] = _public_maybe(entry.value)
    return out


def _public_maybe(value: SconValue | EvalObject) -> SconValue:
    return _public_object(value) if isinstance(value, EvalObject) else value


def _clone_any(value: SconValue | EvalObject) -> SconValue | EvalObject:
    if isinstance(value, EvalObject):
        out = EvalObject()
        for key, entry in value.entries.items():
            out.entries[key] = EvalEntry(_clone_any(entry.value), entry.layer, entry.kind)
        return out
    return clone_value(value)


def _overlay_base(target: EvalObject, source: EvalObject) -> None:
    for key, entry in source.entries.items():
        existing = target.entries.get(key)
        if existing is None:
            target.entries[key] = EvalEntry(_clone_any(entry.value), "base", "ordinary")
        elif existing.layer == "base":
            _overlay_local(existing, _clone_any(entry.value), entry.kind)


def _overlay_local(existing: EvalEntry, value: SconValue | EvalObject, kind: str) -> None:
    if isinstance(existing.value, EvalObject) and isinstance(value, EvalObject):
        _merge_override(existing.value, value)
        existing.layer = "local"
        existing.kind = kind
    else:
        existing.value = value
        existing.layer = "local"
        existing.kind = kind


def _merge_override(target: EvalObject, source: EvalObject) -> None:
    for key, entry in source.entries.items():
        existing = target.entries.get(key)
        if existing is not None and isinstance(existing.value, EvalObject) and isinstance(entry.value, EvalObject):
            _merge_override(existing.value, entry.value)
        else:
            target.entries[key] = EvalEntry(_clone_any(entry.value), entry.layer, entry.kind)


def _invalid_include_path(path: str) -> bool:
    return (
        "://" in path
        or path.startswith("classpath:")
        or "*" in path
        or path.startswith("~")
        or path.startswith("$")
        or os.path.isabs(path)
        or (len(path) >= 3 and path[1] == ":" and path[2] in "\\/")
    )


def _within_root(path: Path, root: Path) -> bool:
    try:
        path.relative_to(root)
        return True
    except ValueError:
        return False


def _read_text_raw(path: Path) -> str:
    with path.open("r", encoding="utf-8", newline="") as file:
        return file.read()


__all__ = ["parse_string", "parse_file", "get_path", "format_value"]

File: go/scon/resolver.go
package scon

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

type LoadOptions struct {
	IncludeRoot string
	Limits      Limits
}

type LoadOption func(*LoadOptions)

type Limits struct {
	MaxFileSize     int
	MaxIncludeDepth int
	MaxIncludeFiles int
	MaxArrayLength  int
	MaxObjectDepth  int
}

func defaultLimits() Limits {
	return Limits{
		MaxFileSize:     16 * 1024 * 1024,
		MaxIncludeDepth: 64,
		MaxIncludeFiles: 1024,
		MaxArrayLength:  1_000_000,
		MaxObjectDepth:  512,
	}
}

func WithIncludeRoot(path string) LoadOption {
	return func(o *LoadOptions) { o.IncludeRoot = path }
}

func WithLimits(limits Limits) LoadOption {
	return func(o *LoadOptions) { o.Limits = limits }
}

func ParseString(source string) (Value, error) {
	doc, err := parseDocument(source, "")
	if err != nil {
		return nil, err
	}
	return newResolver(LoadOptions{Limits: defaultLimits()}).eval(doc)
}

func ParseFile(path string, opts ...LoadOption) (Value, error) {
	options := loadOptions(opts...)
	abs, _ := filepath.Abs(path)
	if options.IncludeRoot == "" {
		options.IncludeRoot = filepath.Dir(abs)
	}
	source, err := os.ReadFile(abs)
	if err != nil {
		return nil, simpleError(IncludeNotFound, "include file not found: "+err.Error())
	}
	if len(source) > options.Limits.MaxFileSize {
		return nil, simpleError(ResourceLimitExceeded, "maximum file size exceeded")
	}
	doc, err := parseDocument(string(source), filepath.Clean(abs))
	if err != nil {
		return nil, err
	}
	resolver := newResolver(options)
	resolver.stack = append(resolver.stack, filepath.Clean(abs))
	resolver.seen[filepath.Clean(abs)] = true
	return resolver.eval(doc)
}

func loadOptions(opts ...LoadOption) LoadOptions {
	options := LoadOptions{Limits: defaultLimits()}
	for _, opt := range opts {
		opt(&options)
	}
	if options.Limits.MaxFileSize == 0 {
		options.Limits = defaultLimits()
	}
	if options.IncludeRoot != "" {
		abs, _ := filepath.Abs(options.IncludeRoot)
		options.IncludeRoot = filepath.Clean(abs)
	}
	return options
}

type layer int
type kind int

const (
	layerBase layer = iota
	layerLocal
)
const (
	kindStructuralObject kind = iota
	kindOrdinaryValue
)

type evalEntry struct {
	value Value
	layer layer
	kind  kind
}

type evalObject struct {
	entries map[string]*evalEntry
	order   []string
}

func (*evalObject) isValue() {}

type resolver struct {
	options LoadOptions
	root    *evalObject
	stack   []string
	seen    map[string]bool
	cache   map[string]*document
	inProg  [][]string
}

func newResolver(options LoadOptions) *resolver {
	return &resolver{
		options: options,
		root:    newEvalObject(),
		seen:    map[string]bool{},
		cache:   map[string]*document{},
		inProg:  [][]string{{}},
	}
}

func (r *resolver) eval(doc *document) (Value, error) {
	if err := r.evalObject(doc.root, nil, doc.file); err != nil {
		return nil, err
	}
	return publicObject(r.root), nil
}

func (r *resolver) evalObject(obj *astObject, path []string, sourcePath string) error {
	if len(path) > r.options.Limits.MaxObjectDepth {
		return sconError(ResourceLimitExceeded, "maximum object depth exceeded", obj.span)
	}
	localSeen := false
	for _, member := range obj.members {
		switch m := member.(type) {
		case astObjectSpread:
			if localSeen {
				return sconError(InvalidSpread, "object spread must appear before local members", m.span)
			}
			target, err := r.lookup(m.sub.path, m.span)
			if err != nil {
				return err
			}
			spread, ok := target.value.(*evalObject)
			if !ok {
				return sconError(TypeMismatch, "object spread target is not an object", m.span)
			}
			obj, err := r.objectAt(path, m.span)
			if err != nil {
				return err
			}
			overlayBase(obj, spread)
		case astInclude:
			included, err := r.loadInclude(sourcePath, m)
			if err != nil {
				return err
			}
			if err := r.evalObject(included.root, path, included.file); err != nil {
				return err
			}
		case astField:
			localSeen = true
			if err := r.evalField(m, path, sourcePath); err != nil {
				return err
			}
		}
	}
	return nil
}

func (r *resolver) evalField(field astField, current []string, sourcePath string) error {
	target := append(append([]string{}, current...), pathNames(field.path)...)
	if obj, ok := field.value.(astObjectValue); ok {
		if err := r.ensureLocalObject(target, field.span); err != nil {
			return err
		}
		r.inProg = append(r.inProg, target)
		err := r.evalObject(obj.object, target, sourcePath)
		r.inProg = r.inProg[:len(r.inProg)-1]
		return err
	}
	value, err := r.evalValue(field.value, sourcePath)
	if err != nil {
		return err
	}
	return r.insertLocalValue(target, value, kindOrdinaryValue, field.span)
}

func (r *resolver) evalValue(value astValue, sourcePath string) (Value, error) {
	switch v := value.(type) {
	case astNull:
		return Null{}, nil
	case astBool:
		return Bool(v.value), nil
	case astNumber:
		n, err := ParseNumber(v.raw)
		if err != nil {
			return nil, sconError(InvalidNumber, err.Error(), v.span)
		}
		return n, nil
	case astString:
		return r.evalString(v)
	case astSubstitution:
		entry, err := r.lookup(v.path, v.span)
		if err != nil {
			return nil, err
		}
		return deepCopy(entry.value), nil
	case astArray:
		out := Array{}
		for _, item := range v.items {
			if len(out) >= r.options.Limits.MaxArrayLength {
				return nil, sconError(ResourceLimitExceeded, "maximum array length exceeded", item.itemSpan())
			}
			switch it := item.(type) {
			case astArrayValue:
				value, err := r.evalValue(it.value, sourcePath)
				if err != nil {
					return nil, err
				}
				out = append(out, value)
			case astArraySpread:
				entry, err := r.lookup(it.sub.path, it.span)
				if err != nil {
					return nil, err
				}
				values, ok := entry.value.(Array)
				if !ok {
					return nil, sconError(TypeMismatch, "array spread target is not an array", it.span)
				}
				for _, item := range values {
					out = append(out, deepCopy(item))
				}
			}
		}
		return out, nil
	case astObjectValue:
		nested := newResolver(r.options)
		nested.stack = r.stack
		nested.seen = r.seen
		nested.cache = r.cache
		if err := nested.evalObject(v.object, nil, sourcePath); err != nil {
			return nil, err
		}
		return publicObject(nested.root), nil
	}
	return nil, simpleError(UnexpectedToken, "unknown value")
}

func (r *resolver) evalString(value astString) (Value, error) {
	if len(value.parts) == 1 {
		if literal, ok := value.parts[0].(stringLiteralPart); ok {
			return String(literal.value), nil
		}
	}
	var out strings.Builder
	for _, part := range value.parts {
		switch p := part.(type) {
		case stringLiteralPart:
			out.WriteString(p.value)
		case stringInterpolationPart:
			entry, err := r.lookup(p.path, p.span)
			if err != nil {
				return nil, err
			}
			switch v := entry.value.(type) {
			case String:
				out.WriteString(string(v))
			case Number:
				out.WriteString(v.String())
			case Bool:
				if bool(v) {
					out.WriteString("true")
				} else {
					out.WriteString("false")
				}
			default:
				return nil, sconError(TypeMismatch, "interpolation requires string, number, or boolean", p.span)
			}
		}
	}
	return String(out.String()), nil
}

func (r *resolver) lookup(path astPath, span Span) (*evalEntry, error) {
	names := pathNames(path)
	for _, prog := range r.inProg {
		if equalPath(names, prog) {
			return nil, sconError(MissingReference, "reference is not completed yet", span)
		}
	}
	current := r.root
	var entry *evalEntry
	for i, name := range names {
		entry = current.entries[name]
		if entry == nil {
			return nil, sconError(MissingReference, "missing reference '"+name+"'", span)
		}
		if i < len(names)-1 {
			object, ok := entry.value.(*evalObject)
			if !ok {
				return nil, sconError(TypeMismatch, "reference path crosses non-object value", span)
			}
			current = object
		}
	}
	return entry, nil
}

func (r *resolver) ensureLocalObject(path []string, span Span) error {
	current := r.root
	for i, name := range path {
		entry := current.entries[name]
		if entry == nil {
			child := newEvalObject()
			current.entries[name] = &evalEntry{value: child, layer: layerLocal, kind: kindStructuralObject}
			current.order = append(current.order, name)
			current = child
			continue
		}
		object, ok := entry.value.(*evalObject)
		if !ok {
			return sconError(PathConflict, "path conflicts with scalar value", span)
		}
		if i == len(path)-1 {
			if entry.layer == layerLocal && entry.kind != kindStructuralObject {
				return sconError(PathConflict, "object field conflicts with ordinary value", span)
			}
			entry.layer = layerLocal
			entry.kind = kindStructuralObject
		}
		current = object
	}
	return nil
}

func (r *resolver) insertLocalValue(path []string, value Value, k kind, span Span) error {
	current := r.root
	for _, name := range path[:len(path)-1] {
		entry := current.entries[name]
		if entry == nil {
			child := newEvalObject()
			current.entries[name] = &evalEntry{value: child, layer: layerLocal, kind: kindStructuralObject}
			current.order = append(current.order, name)
			current = child
			continue
		}
		object, ok := entry.value.(*evalObject)
		if !ok {
			return sconError(PathConflict, "path conflicts with scalar value", span)
		}
		current = object
	}
	leaf := path[len(path)-1]
	existing := current.entries[leaf]
	if existing == nil {
		current.entries[leaf] = &evalEntry{value: value, layer: layerLocal, kind: k}
		current.order = append(current.order, leaf)
		return nil
	}
	if existing.layer == layerBase {
		overlayLocal(existing, value, k)
		return nil
	}
	if existing.kind == kindStructuralObject && k == kindStructuralObject {
		a, aok := existing.value.(*evalObject)
		b, bok := value.(*evalObject)
		if aok && bok {
			return mergeLocalObjects(a, b, span)
		}
	}
	return sconError(DuplicateKey, "duplicate key '"+leaf+"'", span)
}

func (r *resolver) objectAt(path []string, span Span) (*evalObject, error) {
	current := r.root
	for _, name := range path {
		entry := current.entries[name]
		if entry == nil {
			return nil, sconError(PathConflict, "target object does not exist", span)
		}
		object, ok := entry.value.(*evalObject)
		if !ok {
			return nil, sconError(PathConflict, "target path is not an object", span)
		}
		current = object
	}
	return current, nil
}

func (r *resolver) loadInclude(includingFile string, include astInclude) (*document, error) {
	includePath := include.path.value
	if invalidIncludePath(includePath) {
		return nil, sconError(InvalidIncludePath, "invalid include path", include.span)
	}
	includeRoot := r.options.IncludeRoot
	if includeRoot == "" {
		if includingFile == "" {
			return nil, sconError(InvalidIncludePath, "includes require a file context", include.span)
		}
		includeRoot = filepath.Dir(includingFile)
	}
	base := includeRoot
	if includingFile != "" {
		base = filepath.Dir(includingFile)
	}
	candidate, _ := filepath.Abs(filepath.Join(base, includePath))
	candidate = filepath.Clean(candidate)
	rootAbs, _ := filepath.Abs(includeRoot)
	rootAbs = filepath.Clean(rootAbs)
	if !withinRoot(candidate, rootAbs) {
		return nil, sconError(IncludePathDenied, "include path escapes include root", include.span)
	}
	if containsString(r.stack, candidate) {
		return nil, sconError(IncludeCycle, "include cycle: "+candidate, include.span)
	}
	if len(r.stack) >= r.options.Limits.MaxIncludeDepth {
		return nil, sconError(ResourceLimitExceeded, "maximum include depth exceeded", include.span)
	}
	r.seen[candidate] = true
	if len(r.seen) > r.options.Limits.MaxIncludeFiles {
		return nil, sconError(ResourceLimitExceeded, "maximum include file count exceeded", include.span)
	}
	if doc := r.cache[candidate]; doc != nil {
		return doc, nil
	}
	info, err := os.Stat(candidate)
	if err != nil {
		return nil, sconError(IncludeNotFound, "include file not found: "+err.Error(), include.span)
	}
	if info.IsDir() {
		return nil, sconError(IncludeNotFile, "include path is not a file", include.span)
	}
	if info.Size() > int64(r.options.Limits.MaxFileSize) {
		return nil, sconError(ResourceLimitExceeded, "maximum file size exceeded", include.span)
	}
	source, err := os.ReadFile(candidate)
	if err != nil {
		return nil, sconError(IncludeNotFound, "include file not found: "+err.Error(), include.span)
	}
	r.stack = append(r.stack, candidate)
	doc, err := parseDocument(string(source), candidate)
	r.stack = r.stack[:len(r.stack)-1]
	if err != nil {
		if se, ok := err.(*Error); ok {
			code := IncludeParseError
			if se.Code == InvalidRootType {
				code = IncludeRootTypeError
			}
			return nil, &Error{Code: code, Msg: se.Msg, Span: se.Span}
		}
		return nil, err
	}
	r.cache[candidate] = doc
	return doc, nil
}

func newEvalObject() *evalObject {
	return &evalObject{entries: map[string]*evalEntry{}}
}

func publicObject(object *evalObject) *Object {
	out := NewObject()
	for _, key := range object.order {
		if entry := object.entries[key]; entry != nil {
			out.Set(key, publicValue(entry.value))
		}
	}
	return out
}

func publicValue(v Value) Value {
	switch value := v.(type) {
	case *evalObject:
		return publicObject(value)
	case Array:
		out := make(Array, len(value))
		for i, item := range value {
			out[i] = publicValue(item)
		}
		return out
	default:
		return value
	}
}

func pathNames(path astPath) []string {
	out := make([]string, len(path.segments))
	for i, segment := range path.segments {
		out[i] = segment.value
	}
	return out
}

func equalPath(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func overlayBase(target *evalObject, source *evalObject) {
	for _, key := range source.order {
		entry := source.entries[key]
		existing := target.entries[key]
		if existing == nil {
			target.entries[key] = &evalEntry{value: deepCopy(entry.value), layer: layerBase, kind: kindOrdinaryValue}
			target.order = append(target.order, key)
		} else if existing.layer == layerBase {
			if a, ok := existing.value.(*evalObject); ok {
				if b, ok := entry.value.(*evalObject); ok {
					mergeOverride(a, b)
					continue
				}
			}
			existing.value = deepCopy(entry.value)
		}
	}
}

func overlayLocal(existing *evalEntry, value Value, k kind) {
	if a, ok := existing.value.(*evalObject); ok {
		if b, ok := value.(*evalObject); ok {
			mergeOverride(a, b)
			existing.layer = layerLocal
			existing.kind = k
			return
		}
	}
	existing.value = value
	existing.layer = layerLocal
	existing.kind = k
}

func mergeOverride(target, source *evalObject) {
	for _, key := range source.order {
		entry := source.entries[key]
		if existing := target.entries[key]; existing != nil {
			if a, ok := existing.value.(*evalObject); ok {
				if b, ok := entry.value.(*evalObject); ok {
					mergeOverride(a, b)
					continue
				}
			}
			existing.value = deepCopy(entry.value)
			continue
		}
		target.entries[key] = &evalEntry{value: deepCopy(entry.value), layer: entry.layer, kind: entry.kind}
		target.order = append(target.order, key)
	}
}

func mergeLocalObjects(target, source *evalObject, span Span) error {
	for _, key := range source.order {
		entry := source.entries[key]
		if target.entries[key] != nil {
			return sconError(DuplicateKey, "duplicate key '"+key+"'", span)
		}
		target.entries[key] = &evalEntry{value: deepCopy(entry.value), layer: entry.layer, kind: entry.kind}
		target.order = append(target.order, key)
	}
	return nil
}

func invalidIncludePath(path string) bool {
	return strings.Contains(path, "://") ||
		strings.HasPrefix(path, "classpath:") ||
		strings.Contains(path, "*") ||
		strings.HasPrefix(path, "~") ||
		strings.HasPrefix(path, "$") ||
		strings.HasPrefix(path, "/") ||
		looksLikeWindowsAbsolute(path)
}

func looksLikeWindowsAbsolute(path string) bool {
	return len(path) >= 3 && path[1] == ':' && (path[2] == '/' || path[2] == '\\')
}

func withinRoot(path, root string) bool {
	if runtime.GOOS == "windows" {
		path = strings.ToLower(path)
		root = strings.ToLower(root)
	}
	rel, err := filepath.Rel(root, path)
	return err == nil && rel != ".." && !strings.HasPrefix(rel, ".."+string(filepath.Separator))
}

func containsString(values []string, needle string) bool {
	for _, value := range values {
		if value == needle {
			return true
		}
	}
	return false
}
